zh-CN/hello-world.md

```diff
@@ -0,0 +1,510 @@
+- 编写足够的代码以使测试通过
+- 重构
+
+从表面上看，这可能看起来很乏味，但坚持反馈循环非常重要。
+
+它不仅确保你有 *相关的测试*，还可以确保你通过重构测试的安全性来 *设计优秀的软件*。
+
+看到测试失败是一个重要的检查手段，因为它还可以让你看到错误信息。作为一名开发人员，如果测试失败时不能清楚地说明问题所在，那么使用这个代码库可能会非常困难。
+
+通过确保你的测试 *快速* 并设置你的工具，以便运行测试足够简单，你在编写代码时就可以进入流畅的状态。
+
+如果不写测试，你承诺通过运行你的软件来手动检查你的代码，这会打破你的流畅状态，而且你任何时候都无法将自己从这种状态中拯救出来，尤其是从长远来看。
+
+## 继续前进！更多需求
+
+天呐，我们有更多的需求了。 我们现在需要支持第二个参数，指定问候的语言。 如果一种我们不能识别的语言被传进来，就默认为英语。
+
+我们应该确信，我们可以使用 TDD 轻松实现这一功能！
+
+为通过西班牙语的用户编写测试，将其添加到现有套件。
+
+```go
+ t.Run("in Spanish", func(t *testing.T) {
+        got := Hello("Elodie", "Spanish")
+        want := "Hola, Elodie"
+        assertCorrectMessage(t, got, want)
+    })
+```
+
+记住不要作弊！*先写测试*。 当你尝试运行测试时，编译器 *应该* 会出错，因为你用两个参数而不是一个来调用 `Hello`。
+
+```
+./hello_test.go:27:19: too many arguments in call to Hello
+    have (string, string)
+    want (string)
+```
+
+通过向 `Hello` 添加另一个字符串参数来修复编译问题
+
+```go
+func Hello(name string, language string) string {
+    if name == "" {
+        name = "World"
+    }
+    return helloPrefix + name
+}
+```
+
+当你尝试再次运行测试时，它会抱怨在其他测试和 `main.go` 中没有传递足够的参数给 `Hello`
+
+```
+./hello.go:15:19: not enough arguments in call to Hello
+    have (string)
+    want (string, string)
+```
+
+通过传递空字符串来修复它们。现在，除了我们的新场景外，你的所有测试都应该编译并通过
+
+```
+hello_test.go:29: got 'Hola, Elodie' want 'Hello, Elodie'
+```
+
+这里我们可以使用 `if` 检查语言是否是「西班牙语」，如果是就修改信息
+
+```go
+func Hello(name string, language string) string {
+    if name == "" {
+        name = "World"
+    }
+
+    if language == "Spanish" {
+        return "Hola, " + name
+    }
+
+    return helloPrefix + name
+}
+```
+
+测试现在应该通过了。
+
+现在是 *重构* 的时候了。你应该在代码中看出了一些问题，其中有一些重复的「魔术」字符串。 自己尝试重构它，每次更改都要重新运行测试，以确保重构不会破坏任何内容。
+
+```go
+const spanish = "Spanish"
+const helloPrefix = "Hello, "
+const spanishHelloPrefix = "Hola, "
+
+func Hello(name string, language string) string {
+    if name == "" {
+        name = "World"
+    }
+
+    if language == spanish {
+        return spanishHelloPrefix + name
+    }
+
+    return helloPrefix + name
+}
+```
+
+### 法语
+
+- 写一个测试，断言如果你传递 `"French"` 你会得到 `"Bonjour, "`
+- 看到它失败，检查易读的错误消息
+- 在代码中进行最小的合理更改
+
+你可能写了一些看起来大致如此的东西
+
+```go
+func Hello(name string, language string) string {
+    if name == "" {
+        name = "World"
+    }
+
+    if language == spanish {
+        return spanishHelloPrefix + name
+    }
+
+    if language == french {
+        return frenchHelloPrefix + name
+    }
+
+    return helloPrefix + name
+}
+```
+
+### `switch`
+
+当你有很多 `if` 语句检查一个特定的值时，通常使用 `switch` 语句来代替。 如果我们希望稍后添加更多的语言支持，我们可以使用 `switch` 来重构代码，以便更易于阅读和扩展
+
+```go
+func Hello(name string, language string) string {
+    if name == "" {
+        name = "World"
+    }
+
+    prefix := helloPrefix
+
+    switch language {
+    case french:
+        prefix = frenchHelloPrefix
+    case spanish:
+        prefix = spanishHelloPrefix
+    }
+
+    return prefix + name
+}
+```
+
+写一个测试，添加一个用你选择的语言写的问候，你应该看到它是多么简单，去扩展我们 *惊人* 的功能。
+
+### 最后一次重构？
+
+你可能会争辩说，也许我们的功能有点大。对此最简单的重构是将一些功能提取到另一个函数中，并且你已经知道如何声明函数。
```

Comment:
The simplest refactor for this would be to extract out some functionality into another function and **you already know how to declare functions**.

有吧

Reply:
[上游最新版本](https://github.com/quii/learn-go-with-tests/blob/master/hello-world.md#onelastrefactor)中已经没有后半句了，查了一下，在 quii/learn-go-with-tests#11 中移除了。